client-flink/common/src/main/java/org/apache/flink/runtime/metrics/groups/ShuffleMetricGroup.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.metrics.groups;
+
+import static org.apache.flink.runtime.metrics.scope.ShuffleScopeFormat.SCOPE_SHUFFLE_ID;
+import static org.apache.flink.runtime.metrics.scope.ShuffleScopeFormat.fromConfig;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+import java.util.Collections;
+import java.util.Map;
+
+import org.apache.flink.metrics.CharacterFilter;
+import org.apache.flink.metrics.MetricGroup;
+import org.apache.flink.runtime.metrics.dump.ShuffleQueryScopeInfo;
+
+/** Special {@link MetricGroup} representing a Flink runtime Shuffle. */
+public class ShuffleMetricGroup extends ComponentMetricGroup<TaskMetricGroup> {
+
+  /** The shuffle id uniquely identifying the executed shuffle represented by this metrics group. */
+  private final int shuffleId;
+
+  private final ShuffleIOMetricGroup ioMetrics;
+
+  // ------------------------------------------------------------------------
+
+  public ShuffleMetricGroup(MetricGroup taskIOMetricGroup, int shuffleId, String scopeFormat) {
+    this(((TaskIOMetricGroup) taskIOMetricGroup).parentMetricGroup, shuffleId, scopeFormat);
+  }
+
+  public ShuffleMetricGroup(TaskMetricGroup parent, int shuffleId, String scopeFormat) {
+    super(
+        parent.registry,
```

Comment:
The `registry` is protected, therefore `ShuffleMetricGroup` should be defined in `org.apache.flink.runtime.metrics.groups`.